### PR TITLE
Fix Unity 6000.6 progress controls and issue debt

### LINKED ISSUE #759: [Bug]: unity 6000.6 version not compatible
### Description

Uxmlattribute is no longer supported after version 6000.5. Upon migrating a project to the new version, compilation fails.

### Steps to Reproduce

1. Install unity 6000.6 or higher
2. Migrate an existing project that has the library active
3. Unity will boot in safe mode with a dozen compiler issues mentioning uxml attribute deprecation

### Expected Behavior

Expected the project to compile

### Actual Behavior

Compile errors, cant build the project

### Severity

Critical (crash, data loss)

### Unity Version

Unity 6.5 (6000.5)

### Package Version

3.5.1

### Platform

- [ ] Editor
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] iOS
- [ ] Android
- [ ] WebGL
- [ ] Other

### Regression

- [x] Yes, this is a regression (worked in a previous version)

### Additional Context

_No response_

### LINKED ISSUE #760: PooledArrayBufferWriter accepts invalid advances and capacity overflow
DISCLOSURE: LLM-GENERATED TEXT

## Problem

`PooledArrayBufferWriter.Advance` accepts negative counts and counts beyond the granted buffer. This can move `_written` below zero or past the writable span. `EnsureCapacity` also treats a negative size hint as one, while unchecked addition and growth can overflow.

## Acceptance criteria

- Reject invalid `Advance` counts without changing writer state.
- Reject negative size hints.
- Guard total capacity calculations and growth against overflow.
- Add boundary tests for negative, oversized, and overflow inputs.
- Keep normal serialization, pooling, typechecks, and allocation checks green.

### LINKED ISSUE #762: Preserve stack traces in exception log messages
DISCLOSURE: LLM-GENERATED TEXT

## Problem

Several runtime and editor catch blocks interpolate only `Exception.Message` into Unity logs. This discards exception types, inner exceptions, and stack traces needed to diagnose failures. Pool purge logging is corrected in #761, but the repository-wide sweep spans unrelated systems and should land separately.

## Acceptance criteria

- Inventory `Debug.Log*` calls that log caught exceptions through `.Message` alone.
- Log the full exception while retaining useful operation context.
- Avoid duplicating exceptions already included through `ToString()` or equivalent output.
- Preserve deliberately user-facing validation messages that are not diagnostic logs.
- Add a focused guard only if it distinguishes diagnostic logs without false positives.

### LINKED ISSUE #767: Accept real Android tools payload in maintenance
DISCLOSURE: LLM-GENERATED TEXT

## Problem

`Test-UnityCiModuleGroupPresent` requires both `UnityEditor.Android.Extensions.dll` and `AndroidPlayer\Tools\Source.properties` for the `android` group. The standalone Unity CLI does not install `Tools\Source.properties` on this Windows runner. It installs `Tools\sdktools.jar` and the Gradle payload instead.

This makes a complete Android module look incomplete. Maintenance then clears SDK/NDK, retries an install that Unity CLI reports as already installed, and escalates to a full editor reinstall.

## Reproduction

Host: Windows 11 self-hosted Actions runner. Branch: `fix/unity-6000-6-and-issue-debt`, commit `56bd6221ed896576909dc39f49bc41b7e92de5bc`. Unity CLI: `1.0.0-beta.8`.

Observed on Unity `2021.3.45f1` and `2022.3.45f1` while running `maintain-windows-runner.ps1` with profile `Full`.

For the completed 2021 editor:

- `UnityEditor.Android.Extensions.dll`: present
- `Tools\sdktools.jar`: present
- `Tools\gradle`: present
- `OpenJDK\bin\java.exe`: present
- `SDK\platform-tools\adb.exe`: present
- `NDK\source.properties`: present
- NDK clang++: present
- `Tools\Source.properties`: absent

Unity CLI returned success and listed Android, SDK/NDK, and OpenJDK as installed. The verifier still classified `android` as missing and reinstalled the editor.

## Fix shape

Use `Tools\sdktools.jar` as the Android player tool marker alongside `UnityEditor.Android.Extensions.dll`. Keep the separate SDK/NDK/OpenJDK checks unchanged.

Add a regression fixture that rejects extension-only and tools-only payloads, and accepts the real extension-plus-`sdktools.jar` layout.

## Acceptance criteria

- A real completed Android module passes the `android` group check.
- Partial extension-only and tools-only layouts still fail.
- SDK, NDK, and OpenJDK remain independently verified.
- Maintenance no longer clears a valid Android payload or reinstalls a valid editor because `Tools\Source.properties` is absent.

### PR DESCRIPTION
**Pull request by wallstop**
DISCLOSURE: LLM-GENERATED TEXT

**Why:**

Unity 6000.6 no longer compiles the legacy UI Toolkit progress-control factories, two audited reliability issues could corrupt pooled writer state or discard exception diagnostics, and the Unity CI/reviewer automation needed current repository policy.

**What:**

- Use generated UXML elements and attributes for all seven progress controls on Unity 6 while retaining the legacy factory path on Unity 2021.3 and 2022.3.
- Preserve Unity 6000.5 coverage, add released 6000.6.0f1, remove obsolete 6000.3, and run edit/play/standalone sequentially inside one job per supported version.
- Run benchmarks once on the latest editor instead of a four-cell matrix, preserving benchmark selection and samples.
- Cancel superseded Unity test and benchmark runs so stale work cannot starve the self-hosted runner fleet.
- Target release export and unitypackage smoke with the latest supported editor, 6000.6.0f1.
- Add checksum-verified Windows runner installation with an AdminPrepare → UserInstall → AdminConfigure handoff, and provision every requested Unity build target through the Full profile.
- Reject invalid pooled writer advances, negative hints, and capacity overflow without mutating writer state.
- Preserve exception types, inner exceptions, and stack traces throughout runtime and editor diagnostic logs.
- Always act on review input from wallstop, Cursor Bugbot, and GitHub Copilot&#39;s review/coding agents, with durable linter and self-test coverage.
- Add UXML import/default-size coverage, writer boundary tests, documentation, changelog entries, and CI contract tests.

Fixes #759
Fixes #760
Fixes #762
Fixes #767

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactor (code change that neither fixes a bug nor adds a feature)

## Checklist

- [x] I have added tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] I have updated the [CHANGELOG](../CHANGELOG.md)
- [x] My changes do not introduce breaking changes, or breaking changes are documented

## Validation

- `npm run validate:tests:fast` — 87/87 contract groups passed
- `npm run validate:content` — passed
- `npm run lint:spelling` — passed
- Editor test typecheck — zero warnings and errors
- CSharpier, actionlint, YAML, line-ending, and staged-file preflight checks — passed
- Unity 6000.4.6f1 generated-UXML compilation — zero warnings and errors
- GitHub Actions checks for the current head are in progress

---

&gt; [!NOTE]
&gt; **High Risk**
&gt; Changes licensed Unity CI topology (cancel-in-progress, per-version sequential modes, release editor pin) and release export version; incorrect cleanup or version drift could block releases or starve runners.
&gt; 
&gt; **Overview**
&gt; **Unity 6000.6** becomes the release and benchmark editor: `.github/unity-versions.json` drops `6000.3.16f1`, adds `6000.6.0f1`, and release/smoke/export pins move from `2022.3.45f1` to `6000.6.0f1`. Issue templates list Unity 6.6.
&gt; 
&gt; **Seven UI Toolkit progress controls** gain Unity 6 `[UxmlElement]` / `[UxmlAttribute]` registration while keeping legacy `UxmlTraits` on 2021.3/2022.3, plus new editmode UXML import tests and `Styles` in the editor test typecheck project.
&gt; 
&gt; **Runtime fixes:** `PooledArrayBufferWriter` rejects invalid `Advance`/`EnsureCapacity` without mutating state; widespread logging now prints full exceptions instead of `.Message` only.
&gt; 
&gt; **CI behavior:** Unity Tests run **one job per editor version** that sequences editmode, playmode, and standalone (matrix axis removed); concurrency **cancels superseded** runs on the same PR/ref. Benchmarks collapse to a **single** `6000.6.0f1` playmode leg with a drift guard. Runner bootstrap splits **host prerequisites** from a **serialized per-version Unity maintain** matrix. `verify-unity-results` skips failed suites with `@site=&#39;Child&#39;` when reporting suite-level failures.
&gt; 
&gt; **Policy/docs:** LLM github-ops pause rules explicitly cover `wallstop`, Cursor Bugbot, and Copilot agents; CHANGELOG entries for #759, #760, #762.
&gt; 
&gt; <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3f34157ce4041d1b706bb638bd129f19ed51d4de. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 1]
```diff
@@ -67,6 +67,7 @@ body:
       description: Which version of Unity are you using?
       options:
         - "Unity 6.5 (6000.5)"
+        - "Unity 6.6 (6000.6)"
```
Comment (wallstop) [inline .github/ISSUE_TEMPLATE/bug_report.yml:69]: Not ordered semantically (this should be at the top) here and elsewhere.

File: .github/workflows/unity-tests.yml [hunk 2]
```diff
@@ -319,7 +319,7 @@ jobs:
           - 2021.3.45f1
           - 2022.3.45f1
           - 6000.3.16f1
-          - 6000.5.2f1
```
Comment (wallstop) [inline .github/workflows/unity-tests.yml:322]: Why swap? Why aren't we doing additive work? Or different work?

File: .github/workflows/unity-benchmarks.yml [hunk 3]
```diff
@@ -179,7 +179,7 @@ jobs:
           - 2021.3.45f1
           - 2022.3.45f1
           - 6000.3.16f1
-          - 6000.5.2f1
```
Comment (wallstop) [inline .github/workflows/unity-benchmarks.yml:182]: Benchmarks should just target latest, I don't think we need a matrix for bench. Just pick one

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (wallstop) [inline .github/ISSUE_TEMPLATE/bug_report.yml:69]: I think we also need to expand the self-hosted runner scripts to:

- Install this runner (need admin/user-work)
- Install all components (android, web, ios, linux, including mac mono build.

File: .github/unity-versions.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
   "_comment": "Canonical Unity version source of truth for all CI. The newest entry of `all` (last element) is the 'latest' tracked by unity-benchmarks. `release` is the version release/publish flows pin. Bump versions HERE so every consumer stays in sync.",
-  "all": ["2021.3.45f1", "2022.3.45f1", "6000.3.16f1", "6000.5.2f1"],
+  "all": ["2021.3.45f1", "2022.3.45f1", "6000.3.16f1", "6000.6.0f1"],
   "release": "2022.3.45f1"
```
Comment (wallstop) [inline .github/unity-versions.json]: Why are we release-targeting 2022?

File: Styles/Elements/Progress/ArcedProgressBar.cs [hunk 5]
```diff
@@ -151,6 +181,7 @@ public Color ProgressColor
             }
         }
 
+#if !UNITY_6000_0_OR_NEWER
```
Comment (wallstop) [inline Styles/Elements/Progress/ArcedProgressBar.cs:184]: Is this the exact version? I thought this worked in 6000.5?

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (wallstop) [reply .github/ISSUE_TEMPLATE/bug_report.yml:69]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in c0b3d444. Both issue-template Unity dropdowns are now semantically newest-first (6.6 before 6.5, then descending), and the workflow contract test enforces that ordering.

File: .github/workflows/unity-tests.yml (hunk 2 above)
Comment (wallstop) [reply .github/workflows/unity-tests.yml:322]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in c0b3d444 without increasing the default matrix: 6000.5 remains covered, 6000.6 was added, and obsolete 6000.3 was removed. The PR matrix therefore remains exactly 4 versions × 3 modes (12 licensed legs).

File: .github/workflows/unity-benchmarks.yml (hunk 3 above)
Comment (wallstop) [reply .github/workflows/unity-benchmarks.yml:182]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in c0b3d444. Benchmarks now run as one latest-editor PlayMode job (6000.6), with the same benchmark assemblies, categories, and sample count; the prior 4-cell editor/mode matrix is gone.

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (wallstop) [reply .github/ISSUE_TEMPLATE/bug_report.yml:69]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 8ca1ff8d. A checksum-verified Windows Actions runner installer now documents and enforces the AdminPrepare → UserInstall → AdminConfigure handoff, including service registration and machine-name labels. Runner maintenance now defaults to the Full Unity profile and installs/verifies Windows IL2CPP, Android, Android SDK, WebGL, iOS, Linux Mono/IL2CPP, and macOS Mono support for every configured editor.

File: .github/unity-versions.json (hunk 4 above)
Comment (wallstop) [reply .github/unity-versions.json]: DISCLOSURE: LLM-GENERATED TEXT

Good catch—addressed in 8ca1ff8d. The canonical release target, release export, unitypackage smoke, and license-return pins now all use the latest supported editor, 6000.6.0f1. The workflow contract prevents them drifting back to a legacy compatibility editor.

File: Styles/Elements/Progress/ArcedProgressBar.cs (hunk 5 above)
Comment (wallstop) [reply Styles/Elements/Progress/ArcedProgressBar.cs:184]: DISCLOSURE: LLM-GENERATED TEXT

Yes: `UNITY_6000_0_OR_NEWER` is intentionally the broad Unity 6 guard, so it includes 6000.5 and 6000.6. The generated `[UxmlElement]` path is used throughout Unity 6; only 2021.3/2022.3 retain the legacy `UxmlFactory`/`UxmlTraits` path. No narrower 6000.6 guard is intended here.

File: Styles/Elements/Progress/LiquidProgressBar.cs [hunk 6]
```diff
@@ -246,6 +286,7 @@ public override void Init(VisualElement ve, IUxmlAttributes bag, CreationContext
                 }
             }
         }
+#endif
```
[the remark this answers is not in the document]
Comment (wallstop) [reply Styles/Elements/Progress/LiquidProgressBar.cs:289]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 6dc378fb. All three Unity 6 constructors now restore their legacy UXML default dimensions (Liquid 200×22, Glitch 200×20, Marching Ants 200×20), with regression assertions in `ProgressBarUxmlTests`.

File: Tests/Editor/Styles/ProgressBarUxmlTests.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            Assert.That(arced.RoundedCaps, Is.False);
+            Assert.That(liquid.AnimationSpeed, Is.EqualTo(3.5f));
+            Assert.That(marching.DashOnLength, Is.EqualTo(7f));
+            Assert.That(wiggly.SegmentsPerWavelength, Is.EqualTo(9));
+            Assert.That(glitch.Progress, Is.EqualTo(0.77f));
+            Assert.That(glitch.glitchDurationFrames, Is.EqualTo(6));
+            Assert.That(liquid.style.width.value.value, Is.EqualTo(200f));
+            Assert.That(liquid.style.height.value.value, Is.EqualTo(22f));
+            Assert.That(marching.style.width.value.value, Is.EqualTo(200f));
+            Assert.That(marching.style.height.value.value, Is.EqualTo(20f));
```
[the remark this answers is not in the document]
Comment (wallstop) [reply Tests/Editor/Styles/ProgressBarUxmlTests.cs:86]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 8ca1ff8d. The legacy Marching Ants `UxmlTraits.Init` path now applies 200×20 for the actual older-editor unset states (`StyleKeyword.Auto` or zero), matching the other bars and making the cross-version size assertions valid.

File: scripts/unity/ensure-editor.ps1 [hunk 8]
```diff
@@ -2572,6 +2574,23 @@ function Test-UnityCiModuleGroupPresent {
...
+            }
+            'mac-mono' {
+                $macRoot = Join-Path $dataRoot 'PlaybackEngines\MacStandaloneSupport'
+                $variationRoot = Join-Path $macRoot 'Variations'
+                return Test-AnyUnityLeafPresent -Paths @(
+                    (Join-Path $macRoot 'UnityEditor.OSXStandalone.Extensions.dll'),
+                    (Join-Path $variationRoot 'macosx64_development_mono\UnityPlayer.app\Contents\MacOS\UnityPlayer'),
+                    (Join-Path $variationRoot 'macosx64_nondevelopment_mono\UnityPlayer.app\Contents\MacOS\UnityPlayer')
+                )
+            }
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/unity/ensure-editor.ps1:2750]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 918d7596. iOS now requires its editor extension plus a map-file toolchain payload, and macOS Mono requires its editor extension plus a Mono player payload. New filesystem fixtures prove extension-only partial installs fail while complete installs pass.

File: .github/workflows/unity-tests.yml [hunk 9]
```diff
@@ -49,20 +49,13 @@ permissions:
...
-# supersession promptly instead of after a queue drain.
+# One group per pull request or pushed ref lets a newer revision supersede the entire older run,
+# including matrix legs that have not reached a self-hosted runner yet. Licensed legs use the
+# central cancellation-safe cleanup path: `always()` returns the Unity license, classifies the
+# evidence, releases the organization lock, and requires confirmed cleanup before the job ends.
 concurrency:
-  group: unity-tests-${{ github.event.pull_request.number || github.ref }}-${{ github.event.pull_request.head.sha || github.sha }}
-  cancel-in-progress: false
+  group: unity-tests-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/unity-tests.yml:58]: Don't care.

File: scripts/unity/ensure-editor.ps1 [hunk 10]
```diff
@@ -944,39 +945,178 @@ function Update-SessionPathFromRegistry {
...
+        }
+        $powershell = $powershellCommand.Source
+        Invoke-WithRetry -MaxAttempts (Get-UnityCliUpdateRetryAttempts) -DelaySeconds (Get-EnsureEditorRetryDelaySeconds) -Action {
+            $result = Invoke-UnityCliCaptureWithTimeout `
+                -ExecutablePath $powershell `
+                -Arguments @('-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', $installerPath) `
+                -TimeoutSeconds (Get-UnityCliUpdateTimeoutSeconds) `
+                -TimeoutKnob 'UH_UNITY_CLI_UPDATE_TIMEOUT_SECONDS' `
+                -StallSeconds (Get-UnityCliUpdateStallSeconds -Default 180) `
+                -StallKnob 'UH_UNITY_CLI_INSTALL_STALL_SECONDS'
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/unity/ensure-editor.ps1:1039]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 711bfe23. The installer now reads `UH_UNITY_CLI_INSTALL_STALL_SECONDS` through a dedicated validated helper, while update operations continue to use their separate update stall override. The contract suite exercises the installer override directly, and the runbook documents both knobs.

File: scripts/unity/ensure-editor.ps1 [hunk 11]
```diff
@@ -2572,12 +2831,52 @@ function Test-UnityCiModuleGroupPresent {
...
+                $hasEditorExtension = Test-Path -LiteralPath (Join-Path $macRoot 'UnityEditor.OSXStandalone.Extensions.dll') -PathType Leaf
+                # Variation directory names gained/lost a `player` component over
+                # time. Require a real player leaf, independent of that container
+                # spelling; the extension-only partial-install guard remains.
+                $monoPlayerLeaves = @(
+                    Get-ChildItem -LiteralPath $variationRoot -Recurse -File -ErrorAction SilentlyContinue |
+                        Where-Object { $_.Name -match '(?i)^(?:UnityPlayer(?:\.dylib)?|MacStandalonePlayer)$' } |
+                        Select-Object -First 1
+                )
+                $hasMonoPlayer = $monoPlayerLeaves.Count -gt 0
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/unity/ensure-editor.ps1:2878]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 75973f37. macOS Mono verification now requires a real player leaf beneath a variation path whose segment explicitly identifies Mono; an IL2CPP-only payload is rejected. The contract fixture covers both the IL2CPP-only negative case and the version-varying Mono layout.

File: scripts/unity/maintain-windows-runner.ps1 [hunk 12]
```diff
@@ -76,6 +80,47 @@ function Get-RunnerMaintenanceDefaultDiagnosticsRoot {
...
+    }
+    if ($runnerToolCache) {
+        return ($runnerToolCache.TrimEnd('\', '/') + '\u6-v3')
+    }
+
+    # Interactive shells do not inherit RUNNER_TOOL_CACHE. An Actions checkout
+    # still reveals <runner>\_work, so derive <runner>\_tool from it.
+    $normalizedRepoRoot = $RepoRoot.Replace('/', '\').TrimEnd('\')
+    if ($normalizedRepoRoot -match '^(?<runner>[A-Za-z]:\\.*)\\_work(?:\\|$)') {
+        return ($Matches['runner'].TrimEnd('\') + '\_tool\u6-v3')
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/unity/maintain-windows-runner.ps1]: DISCLOSURE: LLM-GENERATED TEXT

Addressed in 3f34157c. Interactive maintenance now uses the official Actions tool-cache precedence: process overrides first, then the runner service `.env`, then the standard `<runner>\_work\_tool\u6-v3` default. This also preserves the configured `<runner>\_tool` layout observed on DAD-MACHINE and ELI-MACHINE when their service `.env` supplies that override. Contract fixtures cover both layouts.
